### PR TITLE
Fixes #6376: Properly catch all error in ncf api

### PR DESCRIPTION
**Pull request by peckpeck**
http://www.rudder-project.org/redmine/issues/6376

### CODE REVIEW

File: api/ncf_api_flask_app/views.py [hunk 1]
```diff
@@ -86,17 +95,12 @@ def get_techniques():
...
     return resp
-  except :
-    errstr = traceback.format_exc()
-    sys.stderr.write(errstr)
-    sys.stderr.flush()
-    error = jsonify ({ "error": "Unknown internal error while fetching techniques", "details": errstr})
-    error.status_code = 500
-    return  error
+  except Exception as e:
+    return format_error(e, "techniques fetch", 500)
```
Comment (VinceMacBuche) [inline api/ncf_api_flask_app/views.py]: could you say "technique fetching", it would be more meaningful

File: api/ncf_api_flask_app/views.py [hunk 2]
```diff
@@ -109,13 +113,9 @@ def get_generic_methods():
...
     return resp
-  except :
-    errstr = traceback.format_exc()
-    sys.stderr.write(errstr)
-    sys.stderr.flush()
-    error = jsonify ({ "error": "Unknown internal error while fetching generic methods", "details": errstr})
-    error.status_code = 500
-    return  error
+  except Exception as e:
+    return format_error(e, "generic methods fetch", 500)
```
Comment (VinceMacBuche) [inline api/ncf_api_flask_app/views.py]: same here

File: api/ncf_api_flask_app/views.py [hunk 3]
```diff
@@ -133,39 +133,23 @@ def create_technique():
...
-    sys.stderr.write(errstr)
-    sys.stderr.flush()
-    error = jsonify ({ "error": "Unknown internal error while writing a technique", "details": errstr})
-    error.status_code = 500
-    return  error
+    ncf.write_technique(technique,path)
+    return jsonify({ "data": technique }), 201
+
+  except Exception as e:
+    return format_error(e, "technique write", 500)
```
Comment (VinceMacBuche) [inline api/ncf_api_flask_app/views.py]: same here

File: tools/ncf.py [hunk 4]
```diff
@@ -135,7 +159,8 @@ def parse_bundlefile_metadata(content, bundle_type):
       res['class_parameter_id'] = res['bundle_args'].index(res['class_parameter'])+1
     except:
       res['class_parameter_id'] = 0
-      raise Exception("The class_parameter name \"" + res['class_parameter'] + "\" does not seem to match any of the bundle's parameters")
+      name = res['bundle_name'] if bundle_name in res else "unknown"
```
Comment (VinceMacBuche) [inline tools/ncf.py]: That will not work, it should be 

     if 'bundle_name' in res else "unknown"

Comment (VinceMacBuche) [inline tools/ncf.py]: Else you get an exception (looking for bundle_name variable

File: tools/ncf.py [hunk 5]
```diff
@@ -144,10 +169,12 @@ def parse_bundlefile_metadata(content, bundle_type):
   expected_tags = tags[bundle_type] + tags["common"]
   if sorted(res.keys()) != sorted(expected_tags):
     missing_keys = [mkey for mkey in expected_tags if mkey not in set(res.keys())]
-    raise Exception("One or more metadata tags not found before the bundle agent declaration (" + ", ".join(missing_keys) + ")")
+    name = res['bundle_name'] if bundle_name in res else "unknown"
```
Comment (VinceMacBuche) [inline tools/ncf.py]: same here

File: tools/ncf.py (hunk 4 above)
Comment (peckpeck) [inline tools/ncf.py]: Good catch :)

**Comment by peckpeck** [comment]
PR updated

File: tools/ncf.py [hunk 6]
```diff
@@ -489,3 +424,114 @@ def generate_technique_content(technique_metadata):
...
+    content = open(file).read()
+    try:
+      metadata = parse_technique_metadata(content)
+      all_metadata[metadata['bundle_name']] = metadata
+
+      if include_methods_calls:
+        method_calls = parse_technique_methods(file)
+        all_metadata[metadata['bundle_name']]['method_calls'] = method_calls
+    except NcfError as e:
+      errors.append(e)
```
Comment (VinceMacBuche) [inline tools/ncf.py]: That would be awesome if the filename can be added:

      error = NcfError("Could not parse Technique file " + file+", cause is: "+ e.message, e.details)
      errors.append(error)

File: tools/ncf.py [hunk 7]
```diff
@@ -489,3 +424,114 @@ def generate_technique_content(technique_metadata):
...
+  filenames = get_all_generic_methods_filenames(alt_path)
+  errors = []
+
+  for file in filenames:
+    content = open(file).read()
+    try:
+      metadata = parse_generic_method_metadata(content)
+      all_metadata[metadata['bundle_name']] = metadata
+    except NcfError as e:
+      errors.append(e)
```
Comment (VinceMacBuche) [inline tools/ncf.py]: Same here ( 'could nor parse generic method')

**Comment by VinceMacBuche** [comment]
Apart the last addition this seems correct!

File: tools/ncf.py (hunk 6 above)
Comment (VinceMacBuche) [inline tools/ncf.py]: Other proposition:

     error= NcfError("Could not parse Technique file '" + file + "', e+message + ", details: ' + e.details)

Comment (peckpeck) [inline tools/ncf.py]: I Moved the caused by code to NcfError __init__

**Comment by peckpeck** [comment]
PR updated

**Comment by VinceMacBuche** [comment]
Goot to me ! need #178 to be reviewed :)

**Pull request merged**